### PR TITLE
[flang][OpenMP] Delayed privatization MLIR lowering support for `distribute`

### PR DESCRIPTION
**Pull request by ergawy**
Starts delayed privatizaiton support for standalone `distribute` directives. Other flavours of `distribute` are still TODO as well as MLIR to LLVM IR lowering.

### CODE REVIEW

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 1]
```diff
@@ -1932,22 +1937,49 @@ static void genStandaloneDistribute(lower::AbstractConverter &converter,
...
+      converter.getFirOpBuilder().getInsertionBlock()->getParentOp());
   mlir::omp::DistributeOperands distributeClauseOps;
   genDistributeClauses(converter, semaCtx, stmtCtx, item->clauses, loc,
                        distributeClauseOps);
 
-  // TODO: Support delayed privatization.
+  // Privatization for a `distribute` directive is done in the `teams` region to
+  // which the directive binds. Therefore, all privatization logic (delayed as
+  // well as early) happens **before** the `distribute` op is generated (i.e.
+  // inside the parent `teams` op).
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:1949]: @skatrak let me know if this does not match how the `teams` -> `distribitue` combined construct should work according to the current dialect design.

**Review by bhandarkar-pranav — APPROVED** [review]
Thank you for the PR, @ergawy. LGTM

**Review by tblah — APPROVED** [review]
LGTM, thanks!

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 2]
```diff
@@ -1936,24 +1935,35 @@ static void genStandaloneDistribute(lower::AbstractConverter &converter,
...
                        distributeClauseOps);
 
-  // TODO: Support delayed privatization.
   DataSharingProcessor dsp(converter, semaCtx, item->clauses, eval,
                            /*shouldCollectPreDeterminedSymbols=*/true,
-                           /*useDelayedPrivatization=*/false, &symTable);
-  dsp.processStep1();
+                           enableDelayedPrivatizationStaging, &symTable);
+  mlir::omp::PrivateClauseOps privateClauseOps;
+  dsp.processStep1(&privateClauseOps);
```
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:1942]: I think we can pass `distributeClauseOps` directly to `dsp.processStep1()`. That way we also avoid the need for calling `setPrivateSymsAttr` and appending to `getPrivateVarsMutable` below.

Comment (ergawy) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:1942]: Ah, right, good idea. Done!

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp [hunk 3]
```diff
@@ -1859,11 +1863,11 @@ LogicalResult SimdOp::verify() {
...
-  // TODO Store clauses in op: privateVars, privateSyms.
-  DistributeOp::build(
-      builder, state, clauses.allocateVars, clauses.allocatorVars,
-      clauses.distScheduleStatic, clauses.distScheduleChunkSize, clauses.order,
-      clauses.orderMod, /*private_vars=*/{}, /*private_syms=*/nullptr);
+  DistributeOp::build(builder, state, clauses.allocateVars,
+                      clauses.allocatorVars, clauses.distScheduleStatic,
+                      clauses.distScheduleChunkSize, clauses.order,
+                      clauses.orderMod, clauses.privateVars,
+                      getPrivateSymsAttrIfPresent(builder, clauses));
```
Comment (skatrak) [inline mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:1870]: I think calling `makeArrayAttr(ctx, clauses.privateSyms)`, as it's done for other builders for operations including the `private` clause, would be better than adding another helper function.

Comment (ergawy) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:1870]: Of course, I did not notice this one, thanks!

**Review by skatrak — APPROVED** [review]
Thank you Kareem, LGTM!

**Pull request merged**